### PR TITLE
strlen microoptimization: use compiletime-constant instead of runtime-constant

### PR DESCRIPTION
**Pull request by ferdymercury**
# This Pull request:

## Changes or fixes:

`strlen` cannot be a constexpr in all platforms otherwise.

This has probably no measurable performance gain, so feel free to close if you deem it's not worth.

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

File: core/foundation/src/TClassEdit.cxx [hunk 1]
```diff
@@ -1419,30 +1419,30 @@ bool TClassEdit::IsStdClass(const char *classname)
 {
    classname += StdLen( classname );
    if ( strcmp(classname,"string")==0 ) return true;
-   if ( strncmp(classname,"bitset<",strlen("bitset<"))==0) return true;
+   if ( strncmp(classname,"bitset<",std::char_traits<char>::length("bitset<"))==0) return true;
```
Comment (guitargeek) [inline core/foundation/src/TClassEdit.cxx:1422]:
```suggestion
   using std::char_traits<char>::length;

   if ( strncmp(classname,"bitset<",length("bitset<"))==0) return true;
```
Personally, I would prefer having a `using` statement at the beginning of `IsStdClass()` to make this less verbose.

**Comment by guitargeek** [comment]
Hi @ferdymercury, thanks for this! It's also better to use the C++ standard library instead of holdovers from C.

> This has probably no measurable performance gain, so feel free to close if you deem it's not worth.

But there is also no disadvantage, right?

**Comment by ferdymercury** [comment]
> But there is also no disadvantage, right?

Not that I know of.

**Comment by guitargeek** [comment]
My bad, I didn't know that was a class member!

Maybe:
```c++
using ct = std::char_traits<char>;
ct::length(...); 
```
?

File: tree/treeplayer/src/TTreePlayer.cxx [hunk 2]
```diff
@@ -2461,43 +2461,43 @@ Long64_t TTreePlayer::Scan(const char *varexp, const char *selection,
 
    if (opt.Contains("lenmax=")) {
       int start = opt.Index("lenmax=");
-      int numpos = start + strlen("lenmax=");
```
Comment (pcanal) [inline tree/treeplayer/src/TTreePlayer.cxx:2464]: This function could also use the shortening of the name (and in general any function where there is more than one use of the function).

Comment (ferdymercury) [reply tree/treeplayer/src/TTreePlayer.cxx:2464]: Changed as suggested

File: core/foundation/src/TClassEdit.cxx [hunk 3]
```diff
@@ -1083,7 +1085,7 @@ int TClassEdit::GetSplit(const char *type, vector<string>& output, int &nestedLo
                offset += 5;
             }
             static const char* char_traits_s = "char_traits<char>";
```
Comment (pcanal) [inline core/foundation/src/TClassEdit.cxx:1087]:
```suggestion
            constexpr auto char_traits_s = "char_traits<char>";
            // or 
            // static constexpr char const* const char_traits_s = "char_traits<char>";
```

**Comment by ferdymercury** [comment]
I guess the test failure is unrelated?

**Comment by guitargeek** [comment]
Yes

**Review by guitargeek — APPROVED** [review]
Thanks a lot!

**Pull request merged**